File: clients/scala/src/main/scala/tyckiting/Ai.scala
package tyckiting

trait Ai {
  def teamName: String
  def makeDecisions(
    roundId: Int,
    events: List[Event],
    bots: List[Bot],
    config: GameConfig): List[Action]
  
  implicit class BotMethods(bot: Bot) {
    def move(x: Int, y: Int) = MoveAction(bot.botId, Position(x, y))
    def radar(x: Int, y: Int) = RadarAction(bot.botId, Position(x, y))
    def cannon(x: Int, y: Int) = CannonAction(bot.botId, Position(x, y))
  }
}

File: clients/javascript/ai/dummy/ai.js
"use strict";

var _ = require("lodash");
var position = require("../../position.js");

// Change botNames and teamName to your choice.
var botNames = [
  "Zero",
  "One",
  "Two"
];

module.exports = function Ai() {
  function makeDecisions(roundId, events, bots, config) {
    bots.forEach(function(bot) {
      var ps = position.neighbours(position.make(bot.x, bot.y), config.move);
      var pos = ps[randInt(0, ps.length - 1)];
      bot.move(pos.x, pos.y);
    });

    _.each(events, function(event) {
      if (event.event === "noaction") {
        console.log("Bot did not respond in required time", event.data);
      }
    });
  }

  function randInt(min, max) {
    var range = max - min;
    var rand = Math.floor(Math.random() * (range + 1));
    return min + rand;
  }

  return {
    // The AI must return these three attributes
    botNames: botNames,
    makeDecisions: makeDecisions
  };
};
